engine/notifier_test.go

```diff
@@ -85,123 +84,112 @@ func TestNotifier_ManyNotifications(t *testing.T) {
 func TestNotifier_ManyConsumers(t *testing.T) {
 	singleTestRun := func(t *testing.T) {
 		t.Parallel()
-		notifier := NewNotifier()
-		c := notifier.Channel()
-
-		// spawn 100 worker routines to each wait for a notification
-		var startingWorkers sync.WaitGroup
-		pendingWorkers := atomic.NewInt32(100)
-		for i := 0; i < 100; i++ {
-			startingWorkers.Add(1)
-			go func() {
-				startingWorkers.Done()
-				<-c
-				pendingWorkers.Dec()
-			}()
-		}
-		startingWorkers.Wait()
-
-		// send 100 notifications, with small delays
-		for i := 0; i < 100; i++ {
-			notifier.Notify()
-			time.Sleep(100 * time.Millisecond)
-		}
+		synctest.Test(t, func(t *testing.T) {
+			notifier := NewNotifier()
+			c := notifier.Channel()
+
+			// spawn 100 worker routines to each wait for a notification
+			pendingWorkers := atomic.NewInt32(100)
+			for range 100 {
+				go func() {
+					<-c
+					pendingWorkers.Dec()
+				}()
+			}
 
-		// require that all workers got a notification
-		if !conditionEventuallySatisfied(func() bool { return pendingWorkers.Load() == 0 }, 3*time.Second, 100*time.Millisecond) {
-			require.Fail(t, "timed out", "still awaiting %d workers to get notification", pendingWorkers.Load())
-		}
+			// wait until all workers are blocked on the notifier channel
+			synctest.Wait()
+
+			for range 100 {
+				notifier.Notify()
+
+				// wait for the previous notification to be consumed to ensure that the producer
+				// won't drop any notifications.
+				synctest.Wait()
+			}
+
+			// wait until all workers are done
+			synctest.Wait()
+
+			// require that all workers got a notification
+			assert.Equal(t, int32(0), pendingWorkers.Load())
+		})
 	}
 
-	for r := 0; r < 100; r++ {
+	for r := range 100 {
 		t.Run(fmt.Sprintf("run %d", r), singleTestRun)
 	}
 }
 
 // TestNotifier_AllWorkProcessed spans many routines pushing work and fewer
-// routines consuming work. We require that all worker is eventually processed.
+// routines consuming work. We require that all work is eventually processed.
 func TestNotifier_AllWorkProcessed(t *testing.T) {
 	singleTestRun := func(t *testing.T) {
 		t.Parallel()
-		notifier := NewNotifier()
-
-		totalWork := int32(100)
-		pendingWorkQueue := make(chan struct{}, totalWork)
-		scheduledWork := atomic.NewInt32(0)
-		consumedWork := atomic.NewInt32(0)
-
-		// starts the consuming first, because if we starts the production first instead, then
-		// we might finish pushing all jobs, before any of our consumer has started listening
-		// to the queue.
-		var consumersAllReady sync.WaitGroup
-		consumersAllReady.Add(5)
-
-		// 5 routines consuming work
-		for i := 0; i < 5; i++ {
-			go func() {
-				consumersAllReady.Done()
-				for consumedWork.Load() < totalWork {
-					<-notifier.Channel()
-				L:
+		synctest.Test(t, func(t *testing.T) {
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
```

Comment:
there is a second call of `cancel()` at the end of the test: https://github.com/onflow/flow-go/blob/0a5f6c7553aac29f94bbd9d92377c1a3db7f3c70/engine/notifier_test.go#L187
should we this one? 
```suggestion
```